### PR TITLE
Add deletion-triggered compaction to RocksJava

### PR DESCRIPTION
**Pull request by rhubner**

### CODE REVIEW

File: java/rocksjni/options.cc [hunk 1]
```diff
@@ -3927,6 +3928,62 @@ jint Java_org_rocksdb_Options_memtableMaxRangeDeletions(JNIEnv*, jobject,
...
+  return retVal;
+}
+
+/*
+ * Class:     org_rocksdb_Options
+ * Method:    setTablePropertiesCollectorFactory
+ * Signature: (J[J)V
+ */
+void Java_org_rocksdb_Options_setTablePropertiesCollectorFactory(
+    JNIEnv* env, jclass, jlong jhandle, jlongArray j_factory_handlers) {
```
Comment (alanpaxton) [inline java/rocksjni/options.cc:3967]: naming - j_factory_handlers or j_factory_handles ?

File: java/src/main/java/org/rocksdb/Options.java [hunk 2]
```diff
@@ -2119,6 +2120,35 @@ public PrepopulateBlobCache prepopulateBlobCache() {
...
+   */
+  public void setTablePropertiesCollectorFactory(List<TablePropertiesCollectorFactory> factories) {
+    long[] factoryHandlers = new long[factories.size()];
+    for (int i = 0; i < factoryHandlers.length; i++) {
+      factoryHandlers[i] = factories.get(i).getNativeHandle();
+    }
+    setTablePropertiesCollectorFactory(nativeHandle_, factoryHandlers);
+  }
+
   //
```
Comment (alanpaxton) [inline java/src/main/java/org/rocksdb/Options.java:2152]: the // END comment is now in the wrong place

File: java/src/main/java/org/rocksdb/Options.java [hunk 3]
```diff
@@ -2119,6 +2120,35 @@ public PrepopulateBlobCache prepopulateBlobCache() {
...
+  public List<TablePropertiesCollectorFactory> tablePropertiesCollectorFactory() {
+    long[] factoryHandlers = tablePropertiesCollectorFactory(nativeHandle_);
+
+    return Arrays.stream(factoryHandlers)
+        .mapToObj(factoryHandle -> TablePropertiesCollectorFactory.newWrapper(factoryHandle))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Set TablePropertiesCollectorFactory in undulation C++ object.
```
Comment (alanpaxton) [inline java/src/main/java/org/rocksdb/Options.java:2139]: "underlying" instead of "undulation"

**Review by alanpaxton — COMMENTED** [review]
There are some minor text/name changes. No fundamental problems. I'm happy with it as the minimal set of functionality which we discussed.

File: java/rocksjni/table_properties_collector_factory.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
+//
+// Created by rhubner on 24-Oct-23.
+//
+
+#include "rocksdb/table_properties.h"
+#include "rocksdb/utilities/table_properties_collectors.h"
+
+#ifndef ROCKSDB_TABLE_PROPERTIES_COLLECTOR_FACTORY_H
```
[the remark this answers is not in the document]
Comment (rhubner) [reply java/rocksjni/table_properties_collector_factory.h:8]: I just followed what was used in other files. And in RocksDBJava use `#pragma once` only in one file.

**Review by ajkr — APPROVED** [review]
LGTM

**Pull request closed without merging**